### PR TITLE
feat: refactor type system and add vector type, support stl index

### PR DESCRIPTION
**Pull request by skyzh**
This pull request refactors the type system and supports more indexes in the system.

* binder,planner: support `using` when creating index. users can choose to create hash table index or btree index.
* index: add stl wrapper into the system. the btree wrapper and hash htable wrapper are now part of the starter code.
* binder: support bind float.
* type: add vector type
* type: refactor the system to represent a type by using `Column`. for example, `VARCHAR(8)` will have type id as varchar and the width is stored in the column struct. expressions and planner have been refactored to take this into account. fixed several bugs related to data width inference.
* add array expressions for vector type. see `vector.slt` on how to use them. i.e., `SELECT ARRAY [1.0, 2.0];`
* tuple: slightly change the tuple format and fix several size computation mistakes.

### CODE REVIEW

**Comment by AveryQi115** [comment]
u r crazy🤣🤣🤣🤣oh my gosh

**Comment by skyzh** [comment]
the vector thing is for my upcoming vector db tutorial... feel free to ask me to drop if it is not ready for you to maintain.

**Comment by AveryQi115** [comment]
> the vector thing is for my upcoming vector db tutorial... feel free to ask me to drop if it is not ready for you to maintain.

No, it is definitely sooooo cool to have vector support in bustub! 🥹

**Comment by ChaosZhai** [comment]
Wanted to further refactor index creation this semester now it's done already 🤯👍

**Comment by yliang412** [comment]
Greatq=! I will take a look later today.

File: src/type/type.cpp [hunk 1]
```diff
@@ -281,13 +292,8 @@ auto Type::GetData(const Value &val __attribute__((unused))) const -> const char
 }
 
 // Get the length of the variable length data
-auto Type::GetLength(const Value &val __attribute__((unused))) const -> uint32_t {
+auto Type::GetStorageSize(const Value &val __attribute__((unused))) const -> uint32_t {
   throw NotImplementedException("GetLength not implemented");
```
Comment (yliang412) [inline src/type/type.cpp:296]:
```suggestion
  throw NotImplementedException("GetStorageSize not implemented");
```

File: test/sql/vector.slt
Comment (yliang412) [inline test/sql/vector.slt:1]: Maybe want to separate the index tests out?

**Review by yliang412 — APPROVED** [review]
LGTM.

File: test/sql/vector.slt
Comment (skyzh) [reply test/sql/vector.slt:1]: yep sounds good!

**Pull request merged**